File: csharp/backend-aspnetcore/CountriesDbClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;

namespace BackendAspNetCore;

public class CountriesDbClient
{
    private const string API_BASE = "https://api.countriesdb.com";
    private const string API_KEY = "YOUR_API_KEY"; // Hardcoded API key
    
    private readonly HttpClient _httpClient;

    public CountriesDbClient()
    {
        _httpClient = new HttpClient
        {
            BaseAddress = new Uri(API_BASE)
        };
        _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", API_KEY);
    }

    public async Task<ValidationResponse> ValidateCountryAsync(Dictionary<string, object> payload)
    {
        return await PostAsync("/api/validate/country", payload);
    }

    public async Task<ValidationResponse> ValidateSubdivisionAsync(Dictionary<string, object> payload)
    {
        return await PostAsync("/api/validate/subdivision", payload);
    }

    public async Task<MultiValidationResponse> ValidateCountriesAsync(Dictionary<string, object> payload)
    {
        return await PostMultiAsync("/api/validate/country", payload);
    }

    public async Task<MultiValidationResponse> ValidateSubdivisionsAsync(Dictionary<string, object> payload)
    {
        return await PostMultiAsync("/api/validate/subdivision", payload);
    }

    private async Task<ValidationResponse> PostAsync(string path, Dictionary<string, object> payload)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync(path, payload);
            var responseContent = await response.Content.ReadAsStringAsync();
            
            if (!response.IsSuccessStatusCode)
            {
                // Try to parse error response
                try
                {
                    var errorResult = JsonSerializer.Deserialize<ValidationResponse>(responseContent);
                    if (errorResult != null)
                    {
                        return errorResult;
                    }
                }
                catch
                {
                    // Fallback if parsing fails
                }
                
                return new ValidationResponse
                {
                    Valid = false,
                    Message = $"Request failed: {response.StatusCode}"
                };
            }
            
            var result = JsonSerializer.Deserialize<ValidationResponse>(responseContent);
            return result ?? new ValidationResponse { Valid = false };
        }
        catch (Exception ex)
        {
            return new ValidationResponse
            {
                Valid = false,
                Message = ex.Message
            };
        }
    }

    private async Task<MultiValidationResponse> PostMultiAsync(string path, Dictionary<string, object> payload)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync(path, payload);
            var responseContent = await response.Content.ReadAsStringAsync();
            
            if (!response.IsSuccessStatusCode)
            {
                return new MultiValidationResponse { Results = null };
            }
            
            var result = JsonSerializer.Deserialize<MultiValidationResponse>(responseContent);
            return result ?? new MultiValidationResponse { Results = null };
        }
        catch
        {
            return new MultiValidationResponse { Results = null };
        }
    }
}

File: java/backend-spring/src/main/java/com/countriesdb/examples/spring/CountriesDbClient.java
package com.countriesdb.examples.spring;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

import java.util.Map;

@Component
public class CountriesDbClient {
    private static final String API_BASE = "https://api.countriesdb.com";
    private static final String API_KEY = "YOUR_API_KEY";
    
    private final RestClient restClient;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public CountriesDbClient() {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.setBearerAuth(API_KEY);

        this.restClient = RestClient.builder()
            .baseUrl(API_BASE)
            .defaultHeaders(httpHeaders -> httpHeaders.addAll(headers))
            .build();
    }

    public Map<String, Object> validateCountry(Map<String, Object> payload) {
        return post("/api/validate/country", payload);
    }

    public Map<String, Object> validateSubdivision(Map<String, Object> payload) {
        return post("/api/validate/subdivision", payload);
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> post(String path, Map<String, Object> payload) {
        try {
            return restClient.post()
                .uri(path)
                .body(payload)
                .retrieve()
                .body(Map.class);
        } catch (org.springframework.web.client.HttpClientErrorException.UnprocessableEntity e) {
            // Handle 422 errors gracefully - parse the error response body
            try {
                String responseBody = e.getResponseBodyAsString();
                if (responseBody != null && !responseBody.isEmpty()) {
                    return objectMapper.readValue(responseBody, Map.class);
                }
            } catch (Exception parseException) {
                // Fallback if parsing fails
            }
            return Map.of(
                "valid", false,
                "message", e.getMessage() != null ? e.getMessage() : "Validation failed"
            );
        } catch (org.springframework.web.client.HttpClientErrorException e) {
            // Handle other 4xx errors
            try {
                String responseBody = e.getResponseBodyAsString();
                if (responseBody != null && !responseBody.isEmpty()) {
                    return objectMapper.readValue(responseBody, Map.class);
                }
            } catch (Exception parseException) {
                // Fallback if parsing fails
            }
            return Map.of(
                "valid", false,
                "message", e.getMessage() != null ? e.getMessage() : "Request failed"
            );
        }
    }
}

